Src/FluentAssertions/DataColumnCollectionAssertionExtensions.cs

```diff
@@ -0,0 +1,236 @@
+﻿using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+
+using FluentAssertions.Collections;
+using FluentAssertions.Common;
+using FluentAssertions.Execution;
+
+namespace FluentAssertions
+{
+    public static class DataColumnCollectionAssertionExtensions
+    {
+        /// <summary>
+        /// Asserts that an object reference refers to the exact same object as another object reference.
+        /// </summary>
+        /// <param name="expected">The expected object</param>
+        /// <param name="because">
+        /// A formatted phrase as is supported by <see cref="string.Format(string,object[])" /> explaining why the assertion
+        /// is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.
+        /// </param>
+        /// <param name="becauseArgs">
+        /// Zero or more objects to format using the placeholders in <paramref name="because"/>.
+        /// </param>
+        public static AndConstraint<GenericCollectionAssertions<DataColumn>> BeSameAs(this GenericCollectionAssertions<DataColumn> assertion, DataColumnCollection expected, string because = "", params object[] becauseArgs)
+        {
+            if (assertion.Subject is NonGenericCollectionWrapper<DataColumnCollection, DataColumn> wrapper)
+            {
+                var actualSubject = wrapper.UnderlyingCollection;
+
+                Execute.Assertion
+                    .UsingLineBreaks
+                    .ForCondition(ReferenceEquals(actualSubject, expected))
+                    .BecauseOf(because, becauseArgs)
+                    .WithDefaultIdentifier("collection")
+                    .FailWith("Expected {context} to refer to {0}{reason}, but found {1}.", expected, actualSubject);
+            }
+            else
+            {
+                Execute.Assertion
```

Comment:
Well, because we're leaning on `GenericCollectionAssertions<TItem>` now, if someone makes their own `List<DataTable>` or what have you then it's going to end up hitting this path. I don't believe it is in fact safe to assume that we're guaranteed to have a `NonGenericCollectionWrapper<TCollection, TItem>` here.

Reply:
It's an extension method that only acts on `GenericCollectionAssertions<DataColumn>`. Unless somebody inherits from that class, it'll never be hit. And if you're really want to be defensive on this, use a guard that throws an `InvalidOperationException`.

Reply:
Am I misunderstanding? If someone makes their own `List<DataColumn>` and then does assertions on it, the `.Should()` call will return a `GenericCollectionAssertions<DataColumn>`, just as it would for a `DataColumnCollection`. Isn't that the case??

I'll see if I can make unit tests for this specific circumstance.

Reply:
See tests named `When_generic_collection_is_tested_against_typed_collection_it_should_fail` in the latest commit.

Reply:
Ah. but in that case, it's a design issue and a more technical exception would be better. Something like `InvalidOperationException`.

Reply:
Is that true? I mean, isn't it conceivable that someone could make a test that ensures that a given method returns an _actual_ `DataTableCollection` and if someone down the road changes it to return an `IEnumerable<DataTable>` (e.g. by applying some LINQ queries to it), they want the test to fail because it isn't the exact object any more? In this case, wouldn't a simple test failure make more sense than an `InvalidOperationException`??

Reply:
No, up to now we've been treating such scenarios as test design errors. Same as passing `NaN` to `BeApproximately`.

Reply:
So `BeSameAs` calls are only well-formed if the type matches?

Reply:
Ohh, but you _can't_ pass 37 in because of the type argument. Mm, got it.

Reply:
Okay, the latest commit makes this change. I did a small refactor of AssertionScope.FailWith so that its failure message formatting can be reused.

Reply:
Okeydoke :-)

Reply:
Oh, so, uh, before these changes, `BeSameAs` would throw an assertion failure if it didn't match, but `NotBeSameAs` wouldn't, because if the type doesn't match, well, then the references definitely don't match. Should I change `NotBeSameAs` to fail the assertion if the subject is of the wrong type?

Reply:
This is now implemented in this way, and the latest two commits, https://github.com/fluentassertions/fluentassertions/commit/0c7c7151a61457815a838631ad3d4e71efbe57a1 and https://github.com/fluentassertions/fluentassertions/commit/0216b3209840ea4669db877925c5eeeac3286e66, show the conversion back to an assertion instead of a direct throw with a reusable `FormatFailureMessage` in `AssertionScope`. The code is very parallel; the change is pretty small.

Reply:
>  Should I change NotBeSameAs to fail the assertion if the subject is of the wrong type?

Yes. Because it's an edge case.